src/IonParserBinaryRaw.ts

```diff
@@ -213,28 +217,40 @@ export class ParserBinaryRaw {
         return LongInt.fromIntBytes(Array.prototype.slice.call(bytes), sign);
     }
 
-    private readUnsignedInt() : number {
-        var v = 0, b;
+    static readUnsignedIntFrom(input: BinarySpan, numberOfBytes: number) {
+        let value = 0, bytesRead = 0, byte;
+        if (numberOfBytes < 1)
+            return 0;
 
-        // they have to tell us the length
-        if (this._len < 1) return 0;
+        while (bytesRead < numberOfBytes) {
+            byte = input.next();
+            bytesRead++;
+            value = value << 8;
+            value = value | (byte & 0xff);
+        }
 
-        // shift in all but the last byte (we've already read the first)
-        while (this._len > 0) {
-            b = this._in.next();
-            this._len--;
-            v = v << 8;
-            v = v | (b & 0xff);
+        // We overflowed
+        if (value < 0) {
+            throw new Error("Attempted to read an unsigned int that was larger than 31 bits."
+                + " Use readUnsignedLongIntFrom instead. UInt size: " + numberOfBytes + ", value: " + value
+            );
         }
+        // Fewer bytes than the required `numberOfBytes` were available in the input
+        if (byte === EOF)
+            return undefined;
```

Comment:
I believe that this should throw an `Exception` rather than returning `undefined`, but I'll leave that for another PR.

The original code accidentally omitted the `return` keyword, so it did nothing at all in this particular error case.

Reply:
Agreed on the Exception. Is there an issue tracking this so it's not lost?

Reply:
I've gone ahead and turned it into an thrown `Error`.